shared-services/src/files/file-service.provider.ts

```diff
@@ -0,0 +1,40 @@
+import { CloudinaryFileService } from "./cloudinary-file.service"
+import { CloudinaryFileUploader } from "./cloudinary-file-uploader"
+import { FileServiceInterface } from "./file-service.interface"
+import { FileConfig, FileServiceTypeEnum } from "./file-config"
+
+export class FileServiceProvider {
+  service: FileServiceInterface
+
+  private configure(): FileConfig {
+    const fileConfig: FileConfig = {
+      fileServiceType: FileServiceTypeEnum.cloudinary,
```

Comment:
How is this going to work when we have many providers? (if there is a design doc section that I'm missing here or something sorry!! feel free to just link)

e.g. will `create()` take an optional argument of fileServiceType? Or will it be an env variable and `configure()` checks that directly? or a third option?

Reply:
Piggybacking on this comment to add some notes.  What I'm describing below is not necessarily what the final state will look like since we may still want to support multiple backends per the design doc (ie still be able to access file in Cloudinary even after moving to S3, and vice versa), but it's a reasonable intermediate step.

Importing the FileServiceProvider should look something like ` import { FileServiceProvider } from "path"`.

When configuring the service(s) initially during the app startup phase, there should be a single call to `FileServiceProvider.configure(config)`.  This will instantiate the actual FileService instances for public and private locations.  You will need to make this a static method to keep from having to instantiate a new class every time.

When making a call to the file service, call `FileServiceProvider.getPublicUploadService()`, or something to that effect.  This will return a `FileServiceInterface` object.  You would then call `putFile()` or whatever method.  The whole thing might look something like:

```
const fileService = FileServiceProvider.getPublicUploadService()
const url = fileService.getDownloadUrlForPhoto(fileId)
```

Not to complicate things, but strictly speaking you could also just expose methods through the module rather than using a `FileServiceProvider` class to hold it.  It's a bit more "Javascript-y" compared to the more "Java-ish" approach outlined above, but otherwise they are functionally identical.

@ifranch Let me know if you'd like for me to go over this in a bit more detail.  Happy to hop on a call or whatever

Reply:
[This section](https://docs.google.com/document/d/17SkUZJGT51TAZuxKZgLHyUmdifrpoborvR8gWVGq_BQ/edit?pli=1#bookmark=id.ij1sdji1xao2) of the doc - written by Chris - covers it. Going off of that and Chris's comments here, I think what makes sense to me is:
1. Add a new environment variable for what file service to use for private and public file services (which uses the same FileServiceTypeEnum values) and then any new environment variables needed for S3 keys, bucket names... 
2. At startup time, call create() which in turn calls configure(). This will read the environment variables and convert that to the FileConfig object. Specifically see which of the services we should use and then also read all of its environment variables like keys and bucket names.
3. Instantiate the service using the FileConfig object. Steps 2 and 3 are really just one step, I separated it here into 2 logical steps based on how we transform the data.
4. Clients will call FileServiceProvider.getPublicUploadService() to get the already instantiated file service.

This is pretty much what Chris said except the logic of transforming env variables -> config object is within the configure()/create() call instead of the provider already being given a config object. If this sounds good, I can make the missing changes in a new PR which are moving the configure()/create() call to app startup time and then making an explicit difference between public and private file services.

Reply:
Thanks all -- this generally LGTM. 

What I would add is that we should do some basic env variable validation/set defaults at build time( ?maybeee next.config.js?), and not just runtime in factory creation logic, would be good.

However some logic to create the config object within create() and configure() sounds good, but my (weakly held) preference is to just make sure that there are some basic building blocks at build time.